dash/_patch.py

```diff
@@ -0,0 +1,119 @@
+def _operation(name, location, **kwargs):
+    return {"operation": name, "location": location, "params": dict(**kwargs)}
+
+
+_noop = object()
+
+
+def validate_slice(obj):
+    if isinstance(obj, slice):
+        raise TypeError("a slice is not a valid index for patch")
+
+
+class Patch:
+    """
+    Patch a callback output value
+
+    Act like a proxy of the output prop value on the frontend.
+
+    Supported prop types: Dictionaries and lists.
+    """
+
+    def __init__(self, location=None, parent=None):
+        if location is not None:
+            self._location = location
+        else:
+            # pylint: disable=consider-using-ternary
+            self._location = (parent and parent._location) or []
+        if parent is not None:
+            self._operations = parent._operations
+        else:
+            self._operations = []
+
+    def __getitem__(self, item):
+        validate_slice(item)
+        return Patch(location=self._location + [item], parent=self)
+
+    def __getattr__(self, item):
+        if item == "tolist":
+            # to_json fix
+            raise AttributeError
+        if item == "_location":
+            return self._location
+        if item == "_operations":
+            return self._operations
+        return self.__getitem__(item)
+
+    def __setattr__(self, key, value):
+        if key in ("_location", "_operations"):
+            self.__dict__[key] = value
+        else:
+            self.__setitem__(key, value)
+
+    def __delattr__(self, item):
+        self.__delitem__(item)
+
+    def __setitem__(self, key, value):
+        validate_slice(key)
+        if value is _noop:
+            # The += set themselves.
+            return
+        self._operations.append(
+            _operation(
+                "Assign",
+                self._location + [key],
+                value=value,
+            )
+        )
+
+    def __delitem__(self, key):
+        validate_slice(key)
+        self._operations.append(_operation("Delete", self._location + [key]))
+
+    def __iadd__(self, other):
+        if isinstance(other, (list, tuple)):
+            self.extend(other)
+        elif isinstance(other, dict):
+            self.merge(other)
```

Comment:
Built-in dicts don't merge with `+`, maybe because they're not purely additive, you can overwrite keys? So we could add `__ior__` if you like but I don't think this belongs in `__iadd__`
```py
>>> a = {1:2}
>>> a += {3:4}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +=: 'dict' and 'dict'
>>> a |= {5:6}
>>> a
{1: 2, 5: 6}
>>> a |= {5:7}
>>> a
{1: 2, 5: 7}

```